testctrl/svc/orch/executor.go

```diff
@@ -26,43 +27,61 @@ type kubeExecutor struct {
 	eventChan <-chan *PodWatchEvent
 	session   *types.Session
 	pcd       podCreateDeleter
+	store     store.Store
 }
 
-func newKubeExecutor(index int, pcd podCreateDeleter, watcher *Watcher) *kubeExecutor {
+func newKubeExecutor(index int, pcd podCreateDeleter, watcher *Watcher, store store.Store) *kubeExecutor {
 	return &kubeExecutor{
 		name:    fmt.Sprintf("%d", index),
 		watcher: watcher,
 		pcd:     pcd,
+		store:   store,
 	}
 }
 
 func (k *kubeExecutor) Execute(session *types.Session) error {
-	var err error
-
 	k.setSession(session)
+	k.writeEvent(types.AcceptEvent, nil, "kubernetes executor %v assigned session %v", k.name, session.Name)
 
-	if err = k.provision(); err != nil {
+	k.writeEvent(types.ProvisionEvent, nil, "started provisioning components for session")
+	err := k.provision()
+	if err != nil {
 		err = fmt.Errorf("failed to provision: %v", err)
 		goto endSession
 	}
 
-	if err = k.monitor(); err != nil {
+	k.writeEvent(types.RunEvent, nil, "all containers appear healthy, monitoring during tests")
+	err = k.monitor()
+	if err != nil {
 		err = fmt.Errorf("failed during test: %v", err)
-		goto endSession
 	}
 
 endSession:
-	if logs, err := k.getDriverLogs(); err == nil {
-		glog.Infof("kubeExecutor[%v]: found logs for component (driver) %v: %s",
-			k.name, k.session.Driver.Name, logs)
+	logs, logErr := k.getDriverLogs()
+	if logErr != nil {
+		logErr = fmt.Errorf("failed to fetch logs from driver: %v", logErr)
 	}
 
-	if err = k.clean(); err != nil {
-		glog.Errorf("kubeExecutor[%v]: failed to teardown resources for session %v: %v",
-			k.name, session.Name, err)
+	cleanErr := k.clean()
+	if cleanErr != nil {
+		cleanErr = fmt.Errorf("failed to tear-down resources: %v", cleanErr)
 	}
 
-	return err
+	if logErr != nil {
+		k.writeEvent(types.InternalErrorEvent, nil, logErr.Error())
+	}
+
+	if cleanErr != nil {
+		k.writeEvent(types.InternalErrorEvent, logs, cleanErr.Error())
+	}
+
+	if err != nil {
+		k.writeEvent(types.ErrorEvent, logs, err.Error())
+		return err
+	}
+
+	k.writeEvent(types.DoneEvent, logs, "session completed, driver container had exit status of 0")
```

Comment:
By considering all error events separately, we paint a more accurate picture of a set of problems. The log issue, inability to clean and any test errors will all be sent as events. The test error is recorded last, since it is most important to return as the latest event from the service.  In the future, I hope a user can fetch all events for a session. This works towards that goal.

Reply:
Is it OK to return only if there is an ErrorEvent but not an InternalErrorEvent? I would expect the condition in line 78 to be something like if logErr != nil || cleanErr != nil || err != nil {... Otherwise the final status will be "done" with no error even if there are internal errors etc.

Reply:
This is intended. A cleanup or log error is an infrastructure failure on our part and it has nothing to do with the user's test. Failure in these indicates that our tool was not able to clean-up resources or fetch logs. Their test may still have published results to BigQuery and succeeded. Right now, it seems problematic since the logs are all we return. In the future, the logs will not have such a high emphasis.

By design, errors for infrastructure events that are unrelated to the user's test are not returned. They are simply recorded for developers. This is why there is a distinction between INTERNAL_ERROR and ERROR. My plan is to intercept them in the future by posting them to [Stackdriver Error Reporting](https://cloud.google.com/error-reporting). This will organize these issues in a dashboard and signal us or gRPC infrastructure devs.